internal/feeder/bastion/bastion_feeder.go

```diff
@@ -0,0 +1,230 @@
+		if sc := status.Code(err); sc == codes.FailedPrecondition {
+			// Invalid proof
+			klog.V(1).Infof("invalid proof: %v", err)
+			w.WriteHeader(http.StatusForbidden)
+			return
+		}
+		if sc := status.Code(err); sc == codes.AlreadyExists {
+			// old checkpoint is smaller than the latest the witness knows about
+			w.Header().Add("Content-Type", "text/x.tlog.size")
+			w.WriteHeader(http.StatusConflict)
+			if _, err := w.Write([]byte(fmt.Sprintf("%d\n", checkpoint.Size))); err != nil {
+				klog.V(1).Infof("Failed to write size response: %v", err)
+			}
+			return
+		}
+	}
+
+	if cpErr != nil {
+		klog.V(1).Infof("invalid checkpoint: %v", cpErr)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if _, err := w.Write([]byte(fmt.Sprintf("— %s %s\n", n.Sigs[0].Name, n.Sigs[0].Base64))); err != nil {
+		klog.V(1).Infof("Failed to write signature response: %v", err)
+	}
+}
+
+// parseBody reads the incoming request and parses into constituent parts.
+//
+// The request body MUST be a sequence of
+// - a previous size line,
+// - zero or more consistency proof lines,
+// - and an empty line,
+// - followed by a [checkpoint][].
+func parseBody(r io.Reader) (uint64, [][]byte, []byte, error) {
+	b := bufio.NewReader(r)
+	sizeLine, _, err := b.ReadLine()
+	if err != nil {
+		klog.Infof("read sizeline: %v", err)
+		return 0, nil, nil, err
+	}
+	var size uint64
+	if n, err := fmt.Sscanf(string(sizeLine), "old %d", &size); err != nil || n != 1 {
+		klog.Infof("scan sizeline: %v", err)
+		return 0, nil, nil, err
+	}
+	proof := [][]byte{}
+	for {
+		l, _, err := b.ReadLine()
+		if err != nil {
+			klog.Infof("read proofline: %v", err)
+			return 0, nil, nil, err
+		}
+		if len(l) == 0 {
+			break
+		}
+		hash, err := base64.StdEncoding.DecodeString(string(l))
+		if err != nil {
+			klog.Infof("base64 proof: %v", err)
+			return 0, nil, nil, err
+		}
+		proof = append(proof, hash)
+	}
+	cp, err := io.ReadAll(b)
+	if err != nil {
+		klog.Infof("read cp: %v", err)
+		return 0, nil, nil, err
+	}
+	return size, proof, cp, nil
+}
+
+func connectAndServe(ctx context.Context, host string, handler http.Handler, key ed25519.PrivateKey) error {
+	t := time.NewTicker(5 * time.Second)
+	defer t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-t.C: // Don't spam bastion in a tight loop in case of error below.
+		}
+
+		cert, err := selfSignedCertificate(key)
+		if err != nil {
+			return err
+		}
+		klog.Infof("Connecting to bastion...")
+		ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+		defer cancel()
+		conn, err := (&tls.Dialer{
+			Config: &tls.Config{
+				Certificates: []tls.Certificate{{
+					Certificate: [][]byte{cert},
+					PrivateKey:  key,
+				}},
+				MinVersion: tls.VersionTLS13,
+				MaxVersion: tls.VersionTLS13,
+				NextProtos: []string{"bastion/0"},
+			},
+		}).DialContext(ctx, "tcp", host)
+		if err != nil {
+			klog.Infof("Failed to connect to bastion: %v", err)
+			time.Sleep(5 * time.Second)
```

Comment:
Why this line _and_ the throttle on the outer connect loop?

Reply:
should've been removed when I added the ticker in the outer loop, removed now.